functests/4_latency/latency.go

```diff
@@ -0,0 +1,228 @@
+		Expect(err).ToNot(HaveOccurred())
+
+		if profile.Spec.CPU.Isolated == nil {
+			Skip(fmt.Sprintf("Skip the oslat test, the profile %q does not have isolated CPUs", profile.Name))
+		}
+
+		isolatedCpus := cpuset.MustParse(string(*profile.Spec.CPU.Isolated))
+		// we require at least two CPUs to run oslat test, because one CPU should be used to run the main oslat thread
+		// we can not use all isolated CPUs, because if reserved and isolated include all node CPUs, and reserved CPUs
+		// do not calculated into the Allocated, at least part of time of one of isolated CPUs will be used to run
+		// other node containers
+		// at least two isolated CPUs to run oslat + one isolated CPU used by other containers on the node = at least 3 isolated CPUs
+		if isolatedCpus.Size() < 3 {
+			Skip(fmt.Sprintf("Skip the oslat test, the profile %q has less than two isolated CPUs", profile.Name))
+		}
+
+		workerRTNodes, err := nodes.GetByLabels(testutils.NodeSelectorLabels)
+		Expect(err).ToNot(HaveOccurred())
+		workerRTNodes, err = nodes.MatchingOptionalSelector(workerRTNodes)
+		Expect(err).ToNot(HaveOccurred(), fmt.Sprintf("error looking for the optional selector: %v", err))
+		Expect(workerRTNodes).ToNot(BeEmpty())
+		workerRTNode = &workerRTNodes[0]
+	})
+
+	AfterEach(func() {
+		if err := testclient.Client.Delete(context.TODO(), oslatPod); err != nil {
+			klog.Error(err)
+		}
+	})
+
+	Context("with the oslat image", func() {
+		It("should succeed", func() {
+			oslatPod = getOslatPod(profile, workerRTNode)
+			err := testclient.Client.Create(context.TODO(), oslatPod)
+			Expect(err).ToNot(HaveOccurred())
+
+			timeout, err := strconv.Atoi(latencyTestRuntime)
+			Expect(err).ToNot(HaveOccurred())
+
+			// give two minutes to download the oslat image
+			err = pods.WaitForPhase(oslatPod, corev1.PodRunning, 2*time.Minute)
+			Expect(err).ToNot(HaveOccurred())
+
+			// add additional 2 minutes to give enough time to the cluster to move the pod to Succeeded phase
+			podTimeout := time.Duration(timeout + 120)
+			err = pods.WaitForPhase(oslatPod, corev1.PodSucceeded, podTimeout*time.Second)
+			Expect(err).ToNot(HaveOccurred())
+
+			cmd := []string{"cat", "/rootfs/var/log/oslat.log"}
+			out, err := nodes.ExecCommandOnNode(cmd, workerRTNode)
+			Expect(err).ToNot(HaveOccurred())
+
+			// verify the maximum latency only when it requested, because this value can be very different
+			// on different systems
+			if maximumLatency == -1 {
+				return
+			}
+
+			maximumRegex, err := regexp.Compile(`Maximum:\t*\s*(.*)\s*\(us\)`)
+			Expect(err).ToNot(HaveOccurred())
+
+			latencies := maximumRegex.FindSubmatch([]byte(out))
+			Expect(maximumLatency).ToNot(BeNil())
+
+			for _, lat := range strings.Split(string(latencies[1]), " ") {
+				if lat == "" {
+					continue
+				}
+
+				curr, err := strconv.Atoi(lat)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(curr < maximumLatency).To(BeTrue())
```

Comment:
I'd call it expected latency (not max)
and allow curr to be within 10% larger or 10% smaller than that.
or even have another env that will define.

I think it is important to check for latency 'improvements' because sometime they may mean something is wrong with the test/env/code

Reply:
I called it this way because it the upper bound, so the test latency should not be above the maximum latency. Regarding 10%, you can set any value for the upper bound when you run the container, so if you want to have bigger expected you can pass 22 instead of 20, but the requirement pretty straight forward say that the latency should be less that 20us.

Reply:
I'm talking about such a test as a way to make sure we don't deteriorate too much. not just to attest that we are in compliance with required latency (which yes, is currently 20us)

Reply:
The problem that we do not check the single CPU, we are also checking the number of CPUs and each CPU can have different latency, for example:
```
Maximum:    16537 16 16 15 13 13 13 12 12 12 12 12 12 12 12 12 12 (us)
```
also, I unsure who should investigate this difference, I prefer currently to concentrate on the 20us and once we will have a requirement it will be pretty easy to change the test.